### ISSUE TITLE
Document Post-Processor

### ISSUE THREAD

**Issue by schroederc** [google/kythe#2913]

**Comment by kamahen** [comment]
I've seen a few references to post-processing (e.g., using Apache Beam), but can't find anything about what kind of post-processing is needed. For example, are there specific Kythe schema edges/nodes that are generated during post-processing (and what are they?)? or are there additional items that aren't documented in the schema?

(My reason for asking is that I'm thinking about a different way of processing sources, which would allow quicker updating in response to source changes, using "just-in-time" computation of some relationships rather than a massive batch recomputation ... but I don't know what kinds of relationships need to be computed over the entire source base.)

**Comment by creachadair** [comment]
> I've seen a few references to post-processing (e.g., using Apache Beam), but can't find anything about what kind of post-processing is needed. For example, are there specific Kythe schema edges/nodes that are generated during post-processing (and what are they?)? or are there additional items that aren't documented in the schema?
> 

This isn't a complete answer, but the goal of post-processing is primarily to denormalize certain relationships that are inefficient to query on the fly. These include, for example:

1. Cross-references (given X—ref—D, find all Y such that Y—ref—D). The indexer can only generate the forward edge of that relation, and without denormalization the inverse requires a complete scan of the graph.
2. Override references (given X—ref/call—F, find the least fixpoint of O(F) = F, (F'—O(F))). The indexer can only generate the forward edges of that relation (in particular upward in the class hierarchy of a language with inheritance), and without denormalization the inverse requires a complete scan of the graph.
3. Decorations (for each @—ref—D in file F, provide names, snippets, and type information so the UI can display references with navigational context). Without denormalization even a modestly-sized file fans out to hundreds of targets.
4. Cross-language joining (given @—ref—T where there exists X—generates—T, patch the reference so that the UI can see @—ref—X). This is similar to (3), but with more steps.

In general, the postprocessor does not synthesize any new relationships, but we found the raw graph cannot satisfy usable latency bounds on common UI queries without additional indexes. Early on in the design of Kythe I speculated we could do the denormalization lazily and only for expensive cases, but I was wrong: In our experiments it turned out that it's not just the long tail that is too costly. Even on a fairly small database of facts, indexing is crucial to query performance.

It's a trade-off of course: How much postprocessing one needs, and of which relationships, depends on what queries need to be answered and their performance constraints. The postprocessing we use now is aimed at very low-latency for UI-based cross-reference queries, and it's possible that other batch-processing scenarios would push different choices.

**Comment by kamahen** [comment]
So, if there were a triples database(*) that had the necessary performance for lookups using either the "from" or the "to" nodes in an edge, most of the post-processing wouldn't be needed? (All of it wouldn't be needed?)

Does there need to be special processing of nodes that have a large number of edges (e.g., `printf`)?

(*) In a triples database, all the tables (which represent edges) have an OID plus two fields (from, to) and are indexed on all fields.
Erlang's mnesia appears to have the necessary scaling characteristics, but I'd need to do more research to verify this. Another possibility is Neo4j. Maybe Yedalog. And there may be others. (Prolog, of course, would work beautifully if all the facts could be kept in-memory.)

**Comment by danielmoy-google** [comment]
@schroederc can comment more definitively, but I would guess it wouldn't be able to replace *all* the postprocessing we currently do internally.  If it was sufficiently fast maybe it could replace a bunch of it.

**Comment by creachadair** [comment]
> So, if there were a triples database(*) that had the necessary performance for lookups using either the "from" or the "to" nodes in an edge, most of the post-processing wouldn't be needed?

In principle, yes (at least "some", maybe even "most). So far none of the triple- or quad-store implementations I've tried comes anywhere close to the performance we would need for the current service. I haven't tried mnesia specifically, but you're right that it would be interesting to do some experimental measurements.

If the goal is to support incremental updates, triple- and quad-stores face the same logistical problem that our custom format does: For performance all the tuples have to be denormalized (permuted), which makes merges expensive. (We tried this with the now-obsolete Pregel system, which used triples, and both the processing _and_ the queries were too slow).

More importantly, though, we still don't have a good way to distinguish deletions from renames within the data. The latter is not a limitation of the storage, but of how we get the data in the first place, and is actually the hard part of incremental updates. It _is_ possible a tuple store might make some of the storage logistics easier, though.

**Comment by kamahen** [comment]
"we still don't have a good way to distinguish deletions from renames within the data."

There might be a solution from how Prolog does reconsult. Basically, it removes all facts/predicates defined with a module before loading new ones. There are some subtleties (which I'd need to research), but in general it seems to work. Note that Prolog reconsult will also transitively reconsult dependencies (imported modules); the information for doing this seems to be missing from the Kythe schema.

(Python has some re-load stuff, but it doesn't seem to do what people want. I'm not sure what the differences are from Prolog; it might be that Python is simply too dynamic or has too complex of an import model. Other languages such as Smalltalk and Lisp also have reload mechanism, but I used them so long ago that I've forgotten how well they worked.)

I'll have to think a bit about this ...

**Comment by creachadair** [comment]
> Basically, it removes all facts/predicates defined with a module before loading new ones.

Yes, and that's more or less what we do now: When building a new index, we re-use existing data for compilation records that do not change (provided compilation for that language is stateless).

This could theoretically be done on at a finer grain, if (a) there were a canonical way to name units of compilation, and (b) every entry was tagged with a provenance as to which compilation units produced it. Both of these are theoretically possible, but neither has turned out to be practical: (a) because [_pace_ Bazel] most build tools don't assign names to compilable units, and (b) because it's a lot of data.

Perhaps more importantly, even if both of those problems _were_ already solved, it's not clear that the reference-counting merge implied by (b) could save any time over the current strategy given the volume of data that would be involved. I don't think we ever tried to do it, because anything that requires additional data per-entry has substantial infrastructure cost. The caching strategy Kythe uses now costs only a small overhead per compilation record beyond the entries themselves.

So, I don't think there are any silver bullets hiding under the couch cushions, but it would still be interesting to experiment with alternative representations.

**Comment by kamahen** [comment]
> When building a new index, we re-use existing data for compilation records that do not change (provided compilation for that language is stateless).

I've worked out a caching scheme for Python that takes into account transitive imports; it wasn't as simple as I had originally hoped  ["I have a problem &mdash; I'll use a cache ... and now I have _two_ problems"], and it doesn't play nicely with Bazel (mainly because of details of Bazel's rules about hermetic inputs and outputs) but works fine with [GNU parallel(1)](https://www.gnu.org/software/parallel/). With some tweaks, it would probably work for other languages.

My recollection from Google is that the post-processing took a fair bit of time (and that was after doing a complete re-processing of all the source(*)). Is there documentation on the "caching strategy Kythe uses now"?

(*) Presumably Bazel could avoid re-processing transitively unchanged source.

**Comment by creachadair** [comment]
> My recollection from Google is that the post-processing took a fair bit of time

Yes, it is quite expensive. We were able to improve it quite a bit, but as far as I know it is still the long pole for producing a new index. My knowledge is at least eight months out of date, though.

> (and that was after doing a complete re-processing of all the source(*)).

Yes. It's not quite a "complete" reprocessing, but in order to discover which compilation records have changed it is necessary to "build", and that is the most expensive part of the operation. Build steps can be cached, but you still have to traverse the dependency graph to verify that the cache is valid. If there were a more efficient way to discover which compilation records have changed (and for each that has changed, what its new inputs are), that cost could in principle be greatly reduced.

Of course, there may be some special cases where rebuilding isn't necessary—but curating custom knowledge of how the build works is generally outside the Kythe team's ambit. The compilation record is the sole interface between the build tool and the rest of the pipeline.

> Is there documentation on the "caching strategy Kythe uses now"?

There was internally, but I don't think we ever published it upstream—the implementation is particular to Google. Roughly speaking, it is a map of entry sets keyed by a hash of the compilation record and the indexer version used to generate it. 

> (*) Presumably Bazel could avoid re-processing transitively unchanged source.

Bazel does cache build results for this purpose. But for a sufficiently large corpus, and with a high enough rate of change, the cache doesn't last long enough. For smaller builds the caching is quite effective, and I expect a lot of open-source projects would require very little cost for rebuilds.

**Comment by kamahen** [comment]
> If there were a more efficient way to discover which compilation records have changed (and for each that has changed, what its new inputs are), that cost could in principle be greatly reduced.

I have come up with a caching scheme that seems to work for Python; and it looks as if it's general enough to work for other languages. Bazel complicates things somewhat, but I think that can be handled.

If anyone is interested, I can write this up. What would be a good way of publishing the design?

**Comment by kamahen** [comment]
> [Updating the graph] could theoretically be done at a finer grain, if (a) there were a canonical way to name units of compilation, and (b) every entry was tagged with a provenance as to which compilation units produced it. 

Don't you need the provenance for loading the facts for a single source file for code browsing? (I presume that all the anchors and their edges are pre-loaded so that mouse-over can be fast)

These seem to be of two kinds of provenance: the easy ones are the {Corpus,Root,Path} nodes plus the edges from them. The difficult ones are the semantic nodes because they don't have a Path; I suppose one way of handling them would be to reference count or garbage collect them. (Another way would be to capture the provenance when the output of an indexer is added to the graph; that would require a separate provenance index ... but maybe that index would be useful for the code browser anyway?)

Are there any other cases where provenance can't be figured out from the Kythe facts alone?

**Comment by kamahen** [comment]
GC won't work for semantic nodes/edges (that is, those don't have a Path); provenance would be needed for them. (Anchor nodes have implicit provenance from their Path; and it's probably not worth GCing intern-ed names.)

Assuming that each semantic node and edge has its own object-id, we'd need to map {Corpus,Root,Path} to a list of node/edges. Would this be a significant space overhead?

This is the kind of data store that I'm thinking of:
```
%% vname("a1",Corpus,Root,Path,"c++").loc/start 4
%% vname("a1",Corpus,Root,Path,"c++").loc/end 7
%% vname("a1",Corpus,Root,Path,"c++").node/kind anchor
%% vname("a1",Corpus,Root,Path,"c++").defines/binding vname("Path.foo",Corpus,Root,"","c++") 
%% vname("Path.foo",Corpus,Root,"","c++").node/kind variable
%%
%% becomes the following facts:

corpus_root_path(666, "Corpus", "Root", "Path.cc", "c++"). % interning Corpus,Root,Path,"c++"
corpus_root_path(999, "Corpus", "Root", "", "c++"). % interning Corpus,Root,"","c++"
vname(123, "a1", 666).          % interning the vname
vname(234, "Path.foo", 999).
string(314159, "variable").     % interning the string "variable"
anchor(123, 4, 7).
edge(123, 'defines/binding', 234).
node_kind(234, 314159).
edge_provenance(666, [123]).
node_provenance(666, [234]).
```

**Comment by creachadair** [comment]
Here is the basic invalidation problem for incremental updates. Although it's difficult to solve, it's not that complex to describe. I've written this here because the topic keeps coming up, and I don't think we have ever documented it clearly.

Leaving aside denormalization, the graph for a build at time t comprises the union of the indexer outputs for a set C of compilation records produced by the build. Or if you like:

G[t] = UNION(C[i] : indexer-output(C[i]))

There are three ways the set C can change:

1. A new compilation record C[new] is added.
2. An existing compilation record C[i] is removed.
3. An existing compilation record C[j] is changed to C'[j].

The first hard problem is to tell which of these conditions has occurred. Intuitively, the idea is that if you can tell which is which, you could add and remove entries in G[t] to compute G[t+1].

A compilation record is a snapshot of a set of source files, dependency files, compiler flags, and environment variables, as they existed during the build. Each record is _identified_ by a digest of those elements, and as those inputs change, the identity of the record changes. To distinguish combinations of (1) and (2) from (3) requires a way to "name" (relate) a family of compilation records over time.

Compilation records are not a first-class entity: They're an intermediate datum we use to capture state information from the edit history of the source. Build tools are complicated, and there is no standard format. We designed compilation records to (hopefully) minimize "irrelevant" changes (where there is no dependency relationship), while accurately detecting changes that affect the meaning of program phrases.

To "name" compilation records requires knowledge from the build environment: For example, in a Bazel workspace, each action is associated with a named "rule" in an explicit dependency graph, so the name of that rule can be used to track the actions for that rule over time. Bazel is unusual in this respect; most build tools choose which code to build together by implicit conventions, e.g., Python's rules about what constitutes a module ("this file") or a package ("all the files touched by this `__init__.py`"). There is no standard.

Because a compilation record is not first-class, its identity can't be used to name the objects defined by the code inside it: The name of an object in the graph is a property of its semantic value in its language, and should not (necessarily) change just because its file or some nearby file was modified. 

The graph does keep track of files, which _are_ first class entities. However, the name of a file is not enough to establish provenance, even in relation to its corpus. A given file may participate in more than one compilation record and contribute different entries to the graph in each (the most prominent example being C++ header files, but it can affect any language). Thus, knowing that an entry came from a particular file isn't enough to tell whether it can be updated or removed: One also needs to know whether it arose from that file _in the context of a particular compilation_ that was added, modified, or removed.

Assuming we solve all that, we (still) aren't quite done: The indexer outputs for two compilation records are not necessarily disjoint. So even if we know that C[j] was changed to form C'[j], we cannot simply compute G[t+1] = (G[t] ∖ indexer-output(C[j])) ∪ indexer-output(C'[j]), because we may in doing so delete some entries that were _also_ provided by some other C[k] overlapping with C[j]. A similar problem occurs when we discover that C[i] was deleted.

The current implementation is conservative, but reasonably easy to understand: Cache the indexer outputs by compilation record, build everything, and recompute the union from whatever came out. This ensures we don't delete anything that isn't really gone, and don't leave behind anything in the graph that was removed from the corpus. The amount of metadata required to do this accurately is small.

In order for an incremental indexing scheme to be competitive—even disregarding the issue of denormalization (though that's also crucial)—it would need to address the problem of how to efficiently distinguish which entries need updating without blowing up the storage requirements. Naming families of compilation records is necessary, but not sufficient, to solve it.

**Comment by kamahen** [comment]
Here's a possible solution to the invalidation problem for incremental updates of the graph, assuming I've understood the problem properly and haven't made any typos.

TL;DR: We can label build rules using fingerprints of its inputs+outputs to track provenance of Kythe facts and to determine what should be deleted when incrementing the graph from a new build.

To simplify things, a graph node/edge "change" is treated as a "remove" + "add" for that node/edge. Also, assume that all pointers to graph nodes/edges are by vname.[^0]

Every Kythe node/edge is marked by its "provenance".  There can be multiple Kythe nodes/edges that differ only in provenance (the graph query mechanism would need to remove duplicates, similar to SQL's "SELECT DISTINCT").

The build system's "rules" and "actions" are assumed to be similar to [Bazel's definition](https://docs.bazel.build/versions/master/skylark/rules.html) — that is, each rule has a set of prerequisites, which are evaluated transitively to determine a set of inputs; the rule also defines actions that generate its outputs. The build system can skip running some rules if it determines that the outputs won't change, but this is merely an optimization.

At time _t_, for all rules _i_:

* `rule[t,i].inputs` —  transitive union of sources and outputs of the rule's prerequisites, as determined by the build system.
* `rule[t,i].actions` — uses `rule[t,i].inputs` to generate `rule[t,i].outputs`.
* `rule[t,i].outputs` — the results of running `[rule[t,i].actions`.
* `rule[t,i].indexer_output` — the set of Kythe nodes/edges  that were created by the indexer, using `rule[t,i].inputs` and `rule[t,i].outputs` (this corresponds to @creachadair's `C[i]`). The details of how this is triggered and run varies by build system (e.g. Bazel's  [extra-actions](https://docs.bazel.build/versions/master/be/extra-actions.html)).
* `rule[t,i].fingerprint` — hash of `rule[t,i].inputs`, `rule[t,i].outputs`, `rule[t,i].actions`.[^1]

The Kythe graph is labeled with provenance information on each node/edge, by labeling all the items in `rule[t,i].indexer_output` with `rule[t,i].fingerprint.` If a Kythe node/edge is generated by multiple rules, it will have multiple labels.

To illustrate how we can use this to do incremental updates to the Kythe graph, consider a complete build at time t0 followed by an incremental build at time t1.[^2]

Create graph `G[t1]` from `G[t0]` (this can be a complete or partial build, as long as the targets are equivalent[^2]):

* (`build[t0]` done at time t0.)
* (`G[t0]` created with nodes/edges, each marked with its corresponding `rule[t0,i].fingerprint`.)
* Compute `FINGERPRINTS[t0]` = set of `rule[t0,i].fingerprint` for *all*  rules _i_ for the targets in `build[t1]`, whether new output was   generated by the build system or not.[^3]
* Do the build (at time t1).
* Compute `FINGERPRINTS[t1]`, similar to `FINGERPRINTS[t0]`.
* Remove labels from all Kythe nodes/edges whose fingerprint is in `FINGERPRINTS[t0]` and not in `FINGERPRINTS[t1]`. Delete nodes/edges that have no labels.
* Add Kythe nodes/edges derived from `build[t1]`, each marked with its corresponding `rule[t1,i].fingerprint`. The build system cannot be relied on to do the minimum actions, so when a Kythe node/edge is added, there should be a check for the node/edge already being in `G` with the same fingerprint.

The fingerprints can be stored as an inverted list (mapping fingerprint → list of nodes/edges) outside the graph because they're only needed when updating the graph.

---------------------

[^0]: The vnames can be "interned" for performance. But we don't want direct pointers because that would complicate the memory management.

[^1]: Some build systems, such as Bazel, have additional information, e.g. the name of the rule. These can be added to the fingerprint, but aren't needed to uniquely identify an action. Conversely, Bazel's name isn't sufficient to uniquely identify an action because the rules might have been changed between two builds.

[^2]: It's difficult to give a formal definition of "equivalent targets" that covers everything, but the idea is that if the build tools haven't changed between t0 and t1, and if the actions haven't changed, then the targets are equivalent. If the targets aren't equivalent, you should do a rebuild of the entire codebase and the build tool should prevent recomputing outputs where not needed.

[^3]: E.g, use the messages from `make --always-make --debug=v --dry-run` ... each target has a series of `Prerequisite` lines, followed by a `Must remake target`, followed by one or more commands, followed by `Successfully remade target` (or `No need to remake target` if `--always-make` isn't specified).

**Comment by shahms** [comment]
Leaving complete provenance information is infeasible for any nodes generated by instantiation of templates in the C++ standard library.

**Comment by kamahen** [comment]
> Leaving complete provenance information is infeasible for any nodes generated by instantiation of templates in the C++ standard library.

Yet another reason for not using C++. ;)

Is it just the C++ standard library templates that cause a problem, or templates in general?

Do C++ `#include`s also mess things up? (As opposed to Java/Python-style imports, which aren't a textual substitution.)
